File: src/main/java/com/httpain/constraints/GroupAssignmentConstraintProvider.java
package com.httpain.constraints;

import org.optaplanner.core.api.score.buildin.hardsoft.HardSoftScore;
import org.optaplanner.core.api.score.stream.Constraint;
import org.optaplanner.core.api.score.stream.ConstraintFactory;
import org.optaplanner.core.api.score.stream.ConstraintProvider;
import org.optaplanner.core.api.score.stream.Joiners;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class GroupAssignmentConstraintProvider implements ConstraintProvider {

    @Override
    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
        return new Constraint[]{
                rolesMustMatch(constraintFactory),
                playersMustNotRepeatInsideGroup(constraintFactory),
                playersMustNotRepeatBetweenGroups(constraintFactory),
//                groupMustHaveAtLeastOnePlayer(constraintFactory),
                shouldMatchKeyRanges(constraintFactory),
                penalizeEmptySlots(constraintFactory)
        };
    }

    private Constraint rolesMustMatch(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
                .filter(group -> (group.getTank() != null && group.getTank().getRole() != Role.Tank)
                        || (group.getHealer() != null && group.getHealer().getRole() != Role.Healer)
                        || group.dps().stream().filter(Objects::nonNull).anyMatch(dps -> dps.getRole() != Role.Dps))
                .penalize(HardSoftScore.ofHard(1000))
                .asConstraint("Roles must match");
    }

    private Constraint playersMustNotRepeatInsideGroup(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
                .filter(group -> group.members().stream().filter(Objects::nonNull).distinct().count() < group.members().stream().filter(Objects::nonNull).count())
                .penalize(HardSoftScore.ofHard(1000))
                .asConstraint("Players must not repeat inside group");
    }

    private Constraint playersMustNotRepeatBetweenGroups(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
                .join(constraintFactory
                        .forEachIncludingNullVars(Group.class))
                .filter((a, b) -> a.getGroupNumber() < b.getGroupNumber() && !groupsHaveDistinctPlayers(a, b))
                .penalize(HardSoftScore.ofHard(1000))
                .asConstraint("Players must not repeat between groups");
    }

    private Constraint groupMustHaveAtLeastOnePlayer(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
                .filter(group -> group.members().stream().allMatch(Objects::isNull))
                .penalize(HardSoftScore.ofHard(1000))
                .asConstraint("No empty groups");
    }

    private Constraint penalizeEmptySlots(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
//                .filter(group -> !group.members().stream().filter(Objects::nonNull).toList().isEmpty() && group.members().contains(null))
                .filter(group -> group.members().contains(null))
                .penalize(HardSoftScore.ofSoft(10), group -> Math.toIntExact(group.members().stream().filter(Objects::isNull).count()))
                .asConstraint("Penalize empty slots");
    }

    private Constraint shouldMatchKeyRanges(ConstraintFactory constraintFactory) {
        return constraintFactory
                .forEachIncludingNullVars(Group.class)
                .filter(group -> group.members().stream().anyMatch(Objects::nonNull) && group.members().stream().filter(Objects::nonNull).map(character -> character.getKeyLevel()).collect(Collectors.toSet()).size() <= 1)
                .reward(HardSoftScore.ofSoft(10000))
                .asConstraint("Key ranges should match");
    }

    private boolean groupsHaveDistinctPlayers(Group a, Group b) {
        List<String> nonNullACharacters = a.members().stream().filter(Objects::nonNull).map(Character::getName).toList();
        List<String> nonNullBCharacters = b.members().stream().filter(Objects::nonNull).map(Character::getName).toList();

        return nonNullACharacters.stream().noneMatch(nonNullBCharacters::contains);
    }

}

File: src/main/kotlin/com/httpain/constraints/GroupAssignmentConstraintProvider.kt
package com.httpain.constraints

import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore
import ai.timefold.solver.core.api.score.stream.Constraint
import ai.timefold.solver.core.api.score.stream.ConstraintFactory
import ai.timefold.solver.core.api.score.stream.ConstraintProvider

class GroupAssignmentConstraintProvider : ConstraintProvider {
    override fun defineConstraints(constraintFactory: ConstraintFactory): Array<Constraint> {
        return arrayOf(
            rolesMustMatch(constraintFactory),
            playersMustNotRepeatInsideGroup(constraintFactory),
            playersMustNotRepeatBetweenGroups(constraintFactory),
            shouldMatchKeyRanges(constraintFactory),
            penalizeEmptySlots(constraintFactory)
        )
    }

    internal fun rolesMustMatch(constraintFactory: ConstraintFactory): Constraint {
        return constraintFactory
            .forEachIncludingNullVars(Group::class.java)
            .filter { group ->
                group.tank != null && group.tank!!.role != Role.Tank
                        || group.healer != null && group.healer!!.role != Role.Healer
                        || group.dps().filterNotNull().any { it.role != Role.Dps }
            }
            .penalize(HardSoftScore.ofHard(Scores.PENALTY_ON_ROLE_MISMATCH))
            .asConstraint("Roles must match")
    }

    internal fun playersMustNotRepeatInsideGroup(constraintFactory: ConstraintFactory): Constraint {
        return constraintFactory
            .forEachIncludingNullVars(Group::class.java)
            .filter { group -> group.members().filterNotNull().hasRepeatingMembers() }
            .penalize(HardSoftScore.ofHard(Scores.PENALTY_ON_DUPLICATE_PLAYERS_IN_GROUP))
            .asConstraint("Players must not repeat inside group");
    }

    internal fun playersMustNotRepeatBetweenGroups(constraintFactory: ConstraintFactory): Constraint {
        return constraintFactory
            .forEachIncludingNullVars(Group::class.java)
            .join(constraintFactory.forEachIncludingNullVars(Group::class.java))
            .filter { it, other -> it.groupNumber < other.groupNumber && !groupsHaveDistinctPlayers(it, other) }
            .penalize(HardSoftScore.ofHard(Scores.PENALTY_ON_DUPLICATE_PLAYERS_BETWEEN_GROUPS))
            .asConstraint("Players must not repeat between groups")
    }

    internal fun penalizeEmptySlots(constraintFactory: ConstraintFactory): Constraint {
        return constraintFactory
            .forEachIncludingNullVars(Group::class.java)
            .filter { it.members().contains(null) }
            .penalize(HardSoftScore.ofSoft(Scores.PENALTY_BASE_ON_EMPTY_SLOT)) { group -> group.members().count { it == null } }
            .asConstraint("Penalize empty slots")
    }

    internal fun shouldMatchKeyRanges(constraintFactory: ConstraintFactory): Constraint {
        return constraintFactory
            .forEachIncludingNullVars(Group::class.java)
            .filter { group -> group.members().filterNotNull().isNotEmpty() && group.members().filterNotNull().map { it.keyLevel }.toSet().size <= 1 }
            .reward(HardSoftScore.ofSoft(Scores.REWARD_ON_KEY_RANGE_MATCH))
            .asConstraint("Key ranges should match")
    }

    private fun groupsHaveDistinctPlayers(a: Group, b: Group): Boolean {
        return (a.members().filterNotNull() + b.members().filterNotNull()).hasRepeatingMembers().not()
    }

    private fun List<Any>.hasRepeatingMembers(): Boolean {
        return this.groupingBy { it }.eachCount().any { it.value > 1 }
    }


}